perf/light.cc

```diff
@@ -0,0 +1,651 @@
+	}
+
+	TupleHolder(const TupleHolder &) = delete;
+	TupleHolder(TupleHolder &&) = delete;
+
+	~TupleHolder()
+	{
+		tuples.clear();
+		delete[] mem;
+		mem = nullptr;
+	}
+
+	unsigned char *mem;
+	std::vector<Tuple> tuples;
+	static RandomBytesGenerator random_generator;
+};
+
+RandomBytesGenerator TupleHolder::random_generator{};
+
+////////////////////////////// Light Definitions ///////////////////////////////////////////////////////////////////////
+
+namespace {
+	bool
+	value_equals(Value_t v1, Value_t v2)
+	{
+		assert(v2.size == v1.size);
+		return std::memcmp(v1.data, v2.data, v1.size) == 0;
+	}
+
+	bool
+	key_equals(Value_t k1, Key_t k2)
+	{
+		Key_t key;
+		memcpy(&key, k1.data, sizeof(key));
+		return key == k2;
+	}
+
+	enum {
+		KB = 1 << 10,
+	};
+
+	static constexpr std::size_t light_extent_size = 16 * KB;
+	static std::size_t extents_count = 0;
+
+	inline void *
+	light_malloc_extend(void *ctx)
+	{
+		std::size_t *p_extents_count = (size_t *) ctx;
+		assert(p_extents_count == &extents_count);
+		++*p_extents_count;
+		return malloc(light_extent_size);
+	}
+
+	inline void
+	light_free_extend(void *ctx, void *p)
+	{
+		size_t *p_extents_count = (size_t *) ctx;
+		assert(p_extents_count == &extents_count);
+		--*p_extents_count;
+		free(p);
+	}
+}; // namespace
+
+#define LIGHT_NAME
+#define LIGHT_DATA_TYPE Value_t
+#define LIGHT_KEY_TYPE Key_t
+#define LIGHT_CMP_ARG_TYPE int
+#define LIGHT_EQUAL(a, b, arg) value_equals(a, b)
+#define LIGHT_EQUAL_KEY(a, b, arg) key_equals(a, b)
+
+#include "salad/light.h"
+
+////////////////////////////// Light Fixture ///////////////////////////////////////////////////////////////////////////
+
+class LightFixture : public benchmark::Fixture {
+public:
+	void
+	SetUp(const benchmark::State &state) noexcept final
+	{
+		light_create(&ht, light_extent_size, light_malloc_extend,
+			     light_free_extend, &extents_count, 0);
+	}
+
+	void
+	TearDown(const benchmark::State &state) noexcept final
+	{
+		light_destroy(&ht);
+	}
+protected:
+	void
+	Fill(std::vector<Tuple>::iterator &&begin, std::vector<Tuple>::iterator &&end) noexcept
+	{
+		std::for_each(begin, end,
+			[this](const Tuple &t){ light_insert(&ht, t.hash, t.raw); });
+	}
+
+	void
+	Erase(std::vector<Tuple>::iterator &&begin, std::vector<Tuple>::iterator &&end) noexcept
+	{
+		std::for_each(begin, end,
+			[this](const Tuple &t){ light_delete_value(&ht, t.hash, t.raw); });
+	}
+
+	struct light_core ht;
+};
+
+// Insert random values into light; no warm up - hashtable is empty at the benchmark start.
+BENCHMARK_DEFINE_F(LightFixture, InsertRandValue)
+(benchmark::State& state)
+{
+	std::size_t insertion_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		// state.range(0) stores current range argument.
+		TupleHolder data(state.range(0));
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			insertion_count++;
+		}
```

Comment:
I see serious problem here.
As I can see, you insert `N = state.range(0)` tuples here, but you do it in cycle `auto s : state`. But the hash table is not recreated each cycle, so you add N and more N and more N..
Actually after `TupleHolder data` is destroyed the hash table hold not-existing tuples..
I think you have to reacreate the hash table each time.

Reply:
Thanks. I've reworked benchmark a bit: now there shouldn't be any leaks & tables are filled properly. Also now benchmark relies on templated fixture